test/test_jit_py3.py

```diff
@@ -241,6 +244,117 @@ def foo():
+                pass
+
+        class FooMod(nn.Module):
+            def one(self, x, y):
+                # type: (Tensor, Tensor) -> Tensor
+                return x + y
+
+            def two(self, x):
+                # type: (Tensor) -> Tensor
+                return 2 * x
+
+            def forward(self, x):
+                # type: (Tensor) -> Tensor
+                return self.one(self.two(x), x)
+
+        class BarMod(nn.Module):
+            def one(self, x, y):
+                # type: (Tensor, Tensor) -> Tensor
+                return x * y
+
+            def two(self, x):
+                # type: (Tensor) -> Tensor
+                return 2 / x
+
+            def forward(self, x):
+                # type: (Tensor) -> Tensor
+                return self.two(self.one(x, x))
+
+        def use_module_interface(mod_list, x):
+            # type: (List[OneTwo], Tensor) -> Tensor
+            return mod_list[0].forward(x) + mod_list[1].forward(x)
+
+        def use_class_interface(mod_list, x):
+            # type: (List[OneTwoClass], Tensor) -> Tensor
+            return mod_list[0].two(x) + mod_list[1].one(x, x)
+
+        scripted_foo_mod = torch.jit.script(FooMod())
+        scripted_bar_mod = torch.jit.script(BarMod())
+        self.checkScript(use_module_interface,
+                         ([scripted_foo_mod, scripted_bar_mod], torch.rand(3, 4),))
+        self.checkScript(use_class_interface,
+                         ([scripted_foo_mod, scripted_bar_mod], torch.rand(3, 4),))
+
+        @torch.jit.script
+        def as_module_interface(x):
+            # type: (OneTwo) -> OneTwo
+            return x
+
+        @torch.jit.script
+        def as_class_interface(x):
+            # type: (OneTwoClass) -> OneTwoClass
+            return x
+
+        @torch.jit.script
+        class Foo(object):
+            def one(self, x, y):
+                # type: (Tensor, Tensor) -> Tensor
+                return x + y
+
+            def two(self, x):
+                # type: (Tensor) -> Tensor
+                return 2 * x
+
+            def forward(self, x):
+                # type: (Tensor) -> Tensor
+                return self.one(self.two(x), x)
+
+        # check class object is not a subtype of module interface
+        with self.assertRaisesRegex(RuntimeError, "ScriptModule class can be subtype of module interface"):
+            as_module_interface(Foo())
+
+        class WrongMod(nn.Module):
+            def two(self, x):
+                # type: (int) -> int
+                return 2 * x
+
+            def forward(self, x):
+                # type: (Tensor) -> Tensor
+                return x + torch.randn(3, self.two(3))
+
+        scripted_wrong_mod = torch.jit.script(WrongMod())
+
+        # wrong module that is not compatible with module interface
+        with self.assertRaisesRegex(RuntimeError, "is not compatible with interface"):
+            as_module_interface(scripted_wrong_mod)
+
+
```

Comment:
We need a test that the is_module nature of an interface lasts through serialization.

Reply:
This has been done in test_interface.cpp